hera_cal/abscal.py

```diff
@@ -981,510 +2636,319 @@ def TT_Phi_gain_arr(self):
-            merged_gains.append(AC.ant_phi_gain)
-            merged_gains = [merge_gains(merged_gains)]
-
-        # collate gains
-        if len(merged_gains) == 0:
-            raise ValueError("abscal_run executed without any calibration arguments set to True")
-        gain_dict = merge_gains(merged_gains)
-        flag_dict = odict(list(map(lambda k: (k, np.zeros((Ntimes, Nfreqs), np.bool)), gain_dict.keys())))
-        gain_pols = AC.gain_pols
-        gain_keys = gain_dict.keys()
-
-        # join gain_flag_dict with flag_dict
-        for k in flag_dict:
-            if k in gain_flag_dict:
-                flag_dict[k] += gain_flag_dict[k]
-
-    # make blank gains if no modelfiles
+def get_all_times_and_lsts(hd, solar_horizon=90.0, unwrap=True):
+    '''Extract all times and lsts from a HERAData object
+
+    Arguments:
+        hd: HERAData object intialized with one ore more uvh5 file's metadata
+        solar_horizon: Solar altitude threshold [degrees]. Times are not returned when the Sun is above this altitude.
+        unwrap: increase all LSTs smaller than the first one by 2pi to avoid phase wrapping
+
+    Returns:
+        all_times: list of times in JD in the file or files
+        all_lsts: LSTs (in radians) corresponding to all_times
+    '''
+    all_times = hd.times
+    all_lsts = hd.lsts
+    if len(hd.filepaths) > 1:  # in this case, it's a dictionary
+        all_times = np.ravel([all_times[f] for f in hd.filepaths])
+        all_lsts = np.ravel([all_lsts[f] for f in hd.filepaths])[np.argsort(all_times)]
+    if unwrap:  # avoid phase wraps 
+        all_lsts[all_lsts < all_lsts[0]] += 2 * np.pi
+        
+    # remove times when sun was too high
+    if solar_horizon < 90.0:
+        lat, lon, alt = hd.telescope_location_lat_lon_alt_degrees
+        solar_alts = utils.get_sun_alt(all_times, latitude=lat, longitude=lon)
+        solar_flagged = solar_alts > solar_horizon
+        return all_times[~solar_flagged], all_lsts[~solar_flagged]
+    else:  # skip this step for speed
+        return all_times, all_lsts
+
+
+def get_d2m_time_map(data_times, data_lsts, model_times, model_lsts):
+    '''Generate a dictionary that maps data times to model times via shared LSTs.
+
+    Arguments:
+        data_times: list of times in the data (in JD)
+        data_lsts: list of corresponding LSTs (in radians)
+        model_times: list of times in the mdoel (in JD)
+        model_lsts: list of corresponing LSTs (in radians)
+
+    Returns:
+        d2m_time_map: dictionary uniqely mapping times in the data to times in the model 
+            that are closest in LST. Each model time maps to at most one data time and 
+            each model time maps to at most one data time. Data times without corresponding
+            model times map to None.
+    '''
+    # first produce a map of indices using the LSTs
+    m2d_ind_map = {}  
+    for dind, dlst in enumerate(data_lsts):
```

Comment:
should consider the effects of LST wrap in this FOR loop

Reply:
Done.